README.md
# ysqlite3
A handy SQLite3 wrapper with additional VFS for encryption or other transforming layers.

## Building
### Compiling and Installing
```sh
git clone --recurse-submodules https://github.com/terrakuh/ysqlite3.git && cd ysqlite3
mkdir build && cd build

cmake ..
cmake --build .

ctest

cmake --build . --target install
```

### CMake
```cmake
find_package(ysqlite3 REQUIRED)

add_executable(main "main.cpp")
target_link_libraries(main ysqlite3::ysqlite3)
```

## Usage
### Simple example
```cpp
#include <ysqlite3/database.hpp>

using namespace ysqlite3;

int main()
{
  Database db{"my.db"};
  db.execute(R"(

  CREATE TABLE IF NOT EXISTS t(
    name text
  );

  INSERT INTO t(name) VALUES('hello'), ('world'), (NULL);

  )");

  auto stmt = db.prepare_statement("SELECT * FROM t WHERE name!=?;");
  results r;
  stmt.bind(1, nullptr);

  while ((r = stmt.step())) {
    for (const auto& column : stmt.columns()) {
      const auto text = r.text(column.c_str());
      std::cout << column << ": " << (!text ? "<null>" : text) << '\n';
    }
  }
}
```

### Custom VFS layers
Here is a simple [ROT13](https://en.wikipedia.org/wiki/ROT13) encryption file layer:

```cpp
#include <iostream>
#include <typeinfo>
#include <ysqlite3/database.hpp>
#include <ysqlite3/vfs/page_transforming_file.hpp>
#include <ysqlite3/vfs/sqlite3_file_wrapper.hpp>
#include <ysqlite3/vfs/sqlite3_vfs_wrapper.hpp>
#include <ysqlite3/vfs/vfs.hpp>

using namespace ysqlite3;

class Rot13_file : public vfs::Page_transforming_file<vfs::SQLite3_file_wrapper>
{
public:
  using Parent = vfs::Page_transforming_file<vfs::SQLite3_file_wrapper>;
  using Parent::Parent;

protected:
  void encode_page(Span<std::uint8_t*> page) override
  {
    for (auto& i : page) {
      i += 13;
    }
  }
  void decode_page(Span<std::uint8_t*> page) override
  {
    for (auto& i : page) {
      i -= 13;
    }
  }
};

int main()
{
  vfs::register_vfs(std::make_shared<vfs::SQLite3_vfs_wrapper<Rot13_file>>(vfs::find_vfs(nullptr), "rot13"),
                    false);

  Database db;
  db.open("test.db", open_flag_readwrite | open_flag_create, "rot13");
  // use db
}
```

### As Extension
```c
sqlite3 db;
// register vfs for the 'real' database
sqlite3_open(":memory:", &db);
sqlite3_enable_load_extension(db, 1);
sqlite3_load_extension(db, "/path/to/ysqlite3_crypt_vfs.ext", "ysqlite3_register_crypt_vfs", 0);
sqlite3_close(db);

sqlite3_open_v2("file:real.db?key=x'0f23eabc39'&cipher=aes-256-gcm", &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_URI, "ysqlite3-crypt");
// automatically initializes the database
sqlite3_enable_load_extension(db, 1);
sqlite3_load_extension(db, "/path/to/ysqlite3_crypt_vfs.ext", "ysqlite3_init_crypt_db", 0);
sqlite3_enable_load_extension(db, 0);
// call the following if you don't want to use the method above
// note: if the db already exists you may need to call 'VACUUM'
int n = 32;
sqlite3_file_control(db, 0, SQLITE_FCNTL_RESERVE_BYTES, &n);
```

## Shell
```sh
ysqlite3 -vfs ysqlite3-crypt encrypted.db
sqlite> .filectrl reserve_bytes 32 #only required if first created
sqlite> PRAGMA cipher="aes-256-gcm";
sqlite> PRAGMA key="r'password123'";
```

## License

[MIT](https://github.com/terrakuh/ysqlite3/blob/master/LICENSE) License.


CHANGELOG.md
# Change Log

## [Unreleased]
### Changed
- Bump SQLite3 version from 3.34.1 to 3.40.0
- Change CMake option `YSQLITE3_SHELL_DISABLE_READLINE` to `YSQLITE3_SHELL_ENABLE_READLINE`

### Fixed
- Removed `noexcept` specifier from throwing constructor

## [0.5.0] - 2021-03-08
### Added
- Additional extensions for shell like JSON, rtree, readline and FTS
- Catch2 as 3rd party submodule
- Capitalize all types
- Upgrade SQLite3 to v3.34.1

### Changed
- Renamve crypt VFS from `ysqlite3-crypt-vfs` to `ysqlite3-crypt`

## [0.4.0] - 2020-10-17
### Added
- File name to VFS file
- File support for version 3
- URI parameter support to crypt file for `key` and `cipher`
- Shell program with `ysqlite3-crypt-vfs` VFS
- Encryption extension

### Fixed
- Unlocking in VFS file

## [0.3.0] - 2020-10-01
### Added
- Re-encryption with the new `PRAGMA`s `crypt_transformation` and `chipher`
- Multiple key format support with `PRAGMA key="r'my raw key'"` or `PRAGMA key="x'070617373776f7264313233'"`

### Changed
- Error code from `generic` to `not_a_database` when decryption of database fails

### Removed
- `PRAGMA plain_key`

## [0.2.0] - 2020-09-29
### Added
- Page transforming file
- Cryptographic functions: `sha1`, `sha2`, `sha3`, `md5` and `digest`
- Crypt VFS for main DB + extension
- Encryption example

### Removed
- Layered VFS

## [0.1.0] - 2020-09-22
### Added
- [ROT13](https://en.wikipedia.org/wiki/ROT13) example

### Changed
- Upgraded SQLite3 to v3.33.0
- Minimum C++ version from `14` to `11`
- All headers and sources into one directory (`ysqlite3`)
- Error reporting with `std::error_code` and `std::system_error` instead of custom exceptions

### Removed
- `backward.hpp` 3rd party
- `microsoft/gsl` 3rd party

[Unreleased]: https://github.com/terrakuh/ysqlite3/compare/v0.5.0...dev
[0.5.0]: https://github.com/terrakuh/ysqlite3/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/terrakuh/ysqlite3/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/terrakuh/ysqlite3/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/terrakuh/ysqlite3/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/terrakuh/ysqlite3/compare/v0.0.0...v0.1.0


src/error.hpp
#ifndef YSQLITE3_ERROR_HPP_
#define YSQLITE3_ERROR_HPP_

#include "sqlite3.h"

#include <system_error>
#include <type_traits>

namespace ysqlite3 {

enum class sqlite3_errcond
{
	success = 0,
	/** Generic error */
	generic = 1,
	/** Internal logic error in SQLite */
	internal = 2,
	/** Access permission denied */
	permission = 3,
	/** Callback routine requested an abort */
	abort = 4,
	/** The database file is locked */
	busy = 5,
	/** A table in the database is locked */
	locked = 6,
	/** A malloc() failed */
	memory = 7,
	/** Attempt to write a readonly database */
	read_only = 8,
	/** Operation terminated by sqlite3_interrupt()*/
	operation_interrupted = 9,
	/** Some kind of disk I/O error occurred */
	io = 10,
	/** The database disk image is malformed */
	corrput = 11,
	/** Unknown opcode in sqlite3_file_control() */
	not_found = 12,
	/** Insertion failed because database is full */
	database_full = 13,
	/** Unable to open the database file */
	bad_database = 14,
	/** Database lock protocol error */
	protocol = 15,
	/** Internal use only */
	empty = 16,
	/** The database schema changed */
	schema_changed = 17,
	/** String or BLOB exceeds size limit */
	value_too_big = 18,
	/** Abort due to constraint violation */
	sql_constraint = 19,
	/** Data type mismatch */
	type_mismatch = 20,
	/** Library used incorrectly */
	library_misuse = 21,
	/** Uses OS features not supported on host */
	no_lfs = 22,
	/** Authorization denied */
	auth_denied = 23,
	/** Not used */
	format = 24,
	/** 2nd parameter to sqlite3_bind out of range */
	out_of_range = 25,
	/** File opened that is not a database file */
	not_a_database = 26,
	/** Notifications from sqlite3_log() */
	notice = 27,
	/** Warnings from sqlite3_log() */
	warning = 28,
};

enum class sqlite3_errc
{
	success = 0,
	/** Generic error */
	generic = 1,
	/** Internal logic error in SQLite */
	internal = 2,
	/** Access permission denied */
	permission = 3,
	/** Callback routine requested an abort */
	abort = 4,
	/** The database file is locked */
	busy = 5,
	/** A table in the database is locked */
	locked = 6,
	/** A malloc() failed */
	memory = 7,
	/** Attempt to write a readonly database */
	read_only = 8,
	/** Operation terminated by sqlite3_interrupt()*/
	operation_interrupted = 9,
	/** Some kind of disk I/O error occurred */
	io = 10,
	/** The database disk image is malformed */
	corrupt = 11,
	/** Unknown opcode in sqlite3_file_control() */
	not_found = 12,
	/** Insertion failed because database is full */
	database_full = 13,
	/** Unable to open the database file */
	bad_database = 14,
	/** Database lock protocol error */
	protocol = 15,
	/** Internal use only */
	empty = 16,
	/** The database schema changed */
	schema_changed = 17,
	/** String or BLOB exceeds size limit */
	value_too_big = 18,
	/** Abort due to constraint violation */
	sql_constraint = 19,
	/** Data type mismatch */
	type_mismatch = 20,
	/** Library used incorrectly */
	library_misuse = 21,
	/** Uses OS features not supported on host */
	no_lfs = 22,
	/** Authorization denied */
	auth_denied = 23,
	/** Not used */
	format = 24,
	/** 2nd parameter to sqlite3_bind out of range */
	out_of_range = 25,
	/** File opened that is not a database file */
	not_a_database = 26,
	/** Notifications from sqlite3_log() */
	notice = 27,
	/** Warnings from sqlite3_log() */
	warning = 28,

	missing_collseq   = (generic | (1 << 8)),
	retry             = (generic | (2 << 8)),
	snapshot          = (generic | (3 << 8)),
	read              = (io | (1 << 8)),
	short_read        = (io | (2 << 8)),
	write             = (io | (3 << 8)),
	fsync             = (io | (4 << 8)),
	dir_fsync         = (io | (5 << 8)),
	truncate          = (io | (6 << 8)),
	fstat             = (io | (7 << 8)),
	unlock            = (io | (8 << 8)),
	rdlock            = (io | (9 << 8)),
	delete            = (io | (10 << 8)),
	blocked           = (io | (11 << 8)),
	nomem             = (io | (12 << 8)),
	access            = (io | (13 << 8)),
	checkreservedlock = (io | (14 << 8)),
	lock              = (io | (15 << 8)),
	close             = (io | (16 << 8)),
	dir_close         = (io | (17 << 8)),
	shmopen           = (io | (18 << 8)),
	shmsize           = (io | (19 << 8)),
	shmlock           = (io | (20 << 8)),
	shmmap            = (io | (21 << 8)),
	seek              = (io | (22 << 8)),
	delete_noent      = (io | (23 << 8)),
	mmap              = (io | (24 << 8)),
	gettemppath       = (io | (25 << 8)),
	convpath          = (io | (26 << 8)),
	vnode             = (io | (27 << 8)),
	auth              = (io | (28 << 8)),
	begin_atomic      = (io | (29 << 8)),
	commit_atomic     = (io | (30 << 8)),
	rollback_atomic   = (io | (31 << 8)),
	data              = (io | (32 << 8)),
	sharedcache       = (locked | (1 << 8)),
	vtab              = (locked | (2 << 8)),
	recovery          = (busy | (1 << 8)),
	snapshot          = (busy | (2 << 8)),
	timeout           = (busy | (3 << 8)),
	notempdir         = (bad_database | (1 << 8)),
	isdir             = (bad_database | (2 << 8)),
	fullpath          = (bad_database | (3 << 8)),
	convpath          = (bad_database | (4 << 8)),
	dirtywal          = (bad_database | (5 << 8)) /* Not Used */,
	symlink           = (bad_database | (6 << 8)),
	vtab              = (corrupt | (1 << 8)),
	sequence          = (corrupt | (2 << 8)),
	index             = (corrupt | (3 << 8)),
	recovery          = (read_only | (1 << 8)),
	cantlock          = (read_only | (2 << 8)),
	rollback          = (read_only | (3 << 8)),
	dbmoved           = (read_only | (4 << 8)),
	cantinit          = (read_only | (5 << 8)),
	directory         = (read_only | (6 << 8)),
	rollback          = (abort | (2 << 8)),
	check             = (sql_constraint | (1 << 8)),
	commithook        = (sql_constraint | (2 << 8)),
	foreignkey        = (sql_constraint | (3 << 8)),
	function          = (sql_constraint | (4 << 8)),
	notnull           = (sql_constraint | (5 << 8)),
	primarykey        = (sql_constraint | (6 << 8)),
	trigger           = (sql_constraint | (7 << 8)),
	unique            = (sql_constraint | (8 << 8)),
	vtab              = (sql_constraint | (9 << 8)),
	rowid             = (sql_constraint | (10 << 8)),
	pinned            = (sql_constraint | (11 << 8)),
	recover_wal       = (notice | (1 << 8)),
	recover_rollback  = (notice | (2 << 8)),
	autoindex         = (warning | (1 << 8)),
	user              = (auth_denied | (1 << 8)),
	load_permanently  = (success | (1 << 8)),
	symlink           = (success | (2 << 8)),
};

inline const std::error_category& sqlite3_category() noexcept
{
	static class : public std::error_category
	{
	public:
		const char* name() const noexcept override
		{
			return "sqlite3";
		}
		std::string message(int ec) const override
		{
			if (const auto str = sqlite3_errstr(ec)) {
				return str;
			}

			return "(unknown SQLite3 error)";
		}
	} category;
	return category;
}

inline std::error_code make_error_code(sqlite3_errc ec) noexcept
{
	return { static_cast<int>(ec), sqlite3_category() };
}

enum class errc
{

	bad_vfs_name = 1,
	database_is_closed
};

inline const std::error_category& ysqlite3_category() noexcept
{
	static class : public std::error_category
	{
	public:
		const char* name() const noexcept override
		{
			return "ysqlite3";
		}
		std::string message(int ec) const override
		{
			switch (static_cast<errc>(ec)) {
			default: return "(unknown error code)";
			}
		}
	} category;
	return category;
}

inline std::error_code make_error_code(errc ec) noexcept
{
	return { static_cast<int>(ec), ysqlite3_category() };
}

} // namespace ysqlite3

namespace std {

template<>
struct is_error_code_enum<ysqlite3::errc> : true_type
{};

template<>
struct is_error_code_enum<ysqlite3::sqlite3_errc> : true_type
{};

} // namespace std

#endif
